silkworm/common/log.hpp

```diff
@@ -27,31 +28,40 @@ enum LogLevels {
     LogTrace, LogDebug, LogInfo, LogWarn, LogError, LogCrit, LogNone
 };
 
-// Log to two output streams - typically the console and a log file.
+// Log to one or two output streams - typically the console and optional log file.
 class Logger {
 public:
-    Logger(std::ostream& o1, std::ostream& o2, LogLevels level=LogNone)
+    Logger(LogLevels level=LogInfo, std::ostream& o1=std::cerr, std::ostream& o2=Logger::null_stream())
     : stream(o1, o2), verbosity(level) {}
 
     int level(LogLevels level) { return verbosity = level; }
     int level() { return verbosity; }
 
-    std::ostream& log(LogLevels level, const char* file, int line){
-    std::time_t time = std::time(nullptr);
-    return stream << LogTags[level] \
-                  << std::put_time(std::gmtime(&time), " %Y-%m-%d %H:%M:%S %Z ") \
-                  << file << ":" << line << "| ";
+    std::ostream& log(LogLevels level, const char* file, int line) {
+        std::time_t time = std::time(nullptr);
+        return stream << LogTags[level]
+                      << std::put_time(std::gmtime(&time), " %Y-%m-%d %H:%M:%S %Z ")
+                      << file << ":" << line << "| ";
     }
+
+    static Logger& default_logger();
+    static std::ostream& null_stream();
+
 private:
     teestream stream;
     LogLevels verbosity;
-    static constexpr char const LogTags[7][6] = { "TRACE", "DEBUG", "INFO ", "WARN ", "ERROR", "CRIT ", "NONE" };
+    static constexpr char const LogTags[7][6] = {
+        "TRACE", "DEBUG", "INFO ", "WARN ", "ERROR", "CRIT ", "NONE"
+    };
 };
 
-// log to whatever logger is in scope
-#define SILKWORM_LOG(level_) \
-    if (logger.level() < (level_)) {} \
-    else logger.log(level_, __FILE__, __LINE__)
+#define SILKWORM_LOG_TO(logger_, level_) \
+    if ((logger_).level() < (level_)) {} \
```

Comment:
This check should be the other way round: logger with ERROR verbosity should show ERROR and CRIT messages and ignore TRACE, DEBUG, INFO, WARN.

Reply:
Easy enough, but not my intuition.  It seems we want to see critical messages for any logging level but LogNone.  Your way means that the more critical the logging level the less verbose the logging, and the less information as to why the critical condition occurred.  Different philosophies I guess.

Reply:
The convention I'm talking about is pretty much the de-facto standard. Boost.Log does the same: https://www.boost.org/doc/libs/1_74_0/libs/log/doc/html/log/tutorial/trivial_filtering.html and https://github.com/boostorg/log/blob/develop/include/boost/log/trivial.hpp#L42

Reply:
OK.  Opposite of what I'm used to, but I'll change.